src/sdk/pynni/nni/compression/torch/speedup/compress_modules.py

```diff
@@ -111,28 +115,48 @@ def replace_conv2d(conv, mask):
     else:
         out_channels_index = mask.output_mask.mask_index[1]
         out_channels = out_channels_index.size()[0]
+
     _logger.debug("replace conv2d with in_channels: %d, out_channels: %d", in_channels, out_channels)
     new_conv = torch.nn.Conv2d(in_channels=in_channels,
                                out_channels=out_channels,
                                kernel_size=conv.kernel_size,
                                stride=conv.stride,
                                padding=conv.padding,
                                dilation=conv.dilation,
-                               groups=1, # currently only support groups is 1
+                               groups=conv.groups,
                                bias=conv.bias is not None,
                                padding_mode=conv.padding_mode)
+
     new_conv.to(conv.weight.device)
     tmp_weight_data = tmp_bias_data = None
+
     if mask.output_mask is not None:
         tmp_weight_data = torch.index_select(conv.weight.data, 0, out_channels_index)
         if conv.bias is not None:
             tmp_bias_data = torch.index_select(conv.bias.data, 0, out_channels_index)
-    # NOTE: does not support group
+    else:
+        tmp_weight_data = conv.weight.data
+    # For the convolutional layers that have more than one group
+    # we need to copy the weight group by group, because the input
+    # channal is also divided into serveral groups and each group
+    # filter may have different input channel indexes.
+    input_step = int(conv.in_channels / conv.groups)
+    filter_step = int(out_channels / conv.groups)
     if mask.input_mask is not None:
-        tmp_weight_data = torch.index_select(conv.weight.data if tmp_weight_data is None else tmp_weight_data,
-                                             1, in_channels_index)
-    assert tmp_weight_data is not None, "Conv2d weight should be updated based on masks"
-    new_conv.weight.data.copy_(tmp_weight_data)
+        for groupid in range(conv.groups):
+            start = groupid * input_step
+            end = (groupid + 1) * input_step
```

Comment:
If the input channel is pruned, are we assuming that the input channels are pruned evenly for each group?   Do we print some warning if the input channels are not pruned evenly?

Reply:
Yes, the `GroupMaskConflict` makes sure that each group is pruned evenly before the masks are passed into the `infer_modules_masks`.  I will add several asserts to check if the input channel mask is legal.